File: cdattg_web_golang/models/eleccion.go
package models

import "time"

// Estados del proceso electoral.
const (
	EleccionEstadoBorrador         = "borrador"
	EleccionEstadoInscripcion      = "inscripcion"
	EleccionEstadoVotacion         = "votacion"
	EleccionEstadoEmpatePendiente  = "empate_pendiente"
	EleccionEstadoCerrada          = "cerrada"
)

// Estados de plancha.
const (
	PlanchaEstadoPendiente  = "pendiente_confirmacion"
	PlanchaEstadoConfirmada = "confirmada"
	PlanchaEstadoRechazada  = "rechazada"
	PlanchaEstadoRetirada   = "retirada"
)

// EleccionProceso ciclo electoral anual por regional.
type EleccionProceso struct {
	UserAuditModel
	RegionalID             uint       `gorm:"column:regional_id;not null;uniqueIndex:uk_eleccion_regional_anio" json:"regional_id"`
	Anio                   int        `gorm:"column:anio;not null;uniqueIndex:uk_eleccion_regional_anio" json:"anio"`
	NombreCiclo            string     `gorm:"column:nombre_ciclo;size:120;not null" json:"nombre_ciclo"`
	Estado                 string     `gorm:"column:estado;size:32;not null;default:borrador" json:"estado"`
	FechaInscripcionInicio *time.Time `gorm:"column:fecha_inscripcion_inicio" json:"fecha_inscripcion_inicio,omitempty"`
	FechaInscripcionFin    *time.Time `gorm:"column:fecha_inscripcion_fin" json:"fecha_inscripcion_fin,omitempty"`
	FechaVotacionInicio    *time.Time `gorm:"column:fecha_votacion_inicio" json:"fecha_votacion_inicio,omitempty"`
	FechaVotacionFin       *time.Time `gorm:"column:fecha_votacion_fin" json:"fecha_votacion_fin,omitempty"`
	MinDiasMatricula       *int       `gorm:"column:min_dias_matricula" json:"min_dias_matricula,omitempty"`

	Regional *Regional `gorm:"foreignKey:RegionalID" json:"regional,omitempty"`
}

func (EleccionProceso) TableName() string { return "eleccion_procesos" }

// EleccionPlancha titular + suplente postulados juntos.
type EleccionPlancha struct {
	UserAuditModel
	ProcesoID            uint       `gorm:"column:proceso_id;not null;index" json:"proceso_id"`
	TitularAprendizID    uint       `gorm:"column:titular_aprendiz_id;not null" json:"titular_aprendiz_id"`
	SuplenteAprendizID   uint       `gorm:"column:suplente_aprendiz_id;not null" json:"suplente_aprendiz_id"`
	Estado               string     `gorm:"column:estado;size:32;not null" json:"estado"`
	TitularConfirmadoAt  *time.Time `gorm:"column:titular_confirmado_at" json:"titular_confirmado_at,omitempty"`
	SuplenteConfirmadoAt *time.Time `gorm:"column:suplente_confirmado_at" json:"suplente_confirmado_at,omitempty"`
	PropuestaPorUserID   *uint      `gorm:"column:propuesta_por_user_id" json:"propuesta_por_user_id,omitempty"`
	MotivoRechazo        *string    `gorm:"column:motivo_rechazo;size:500" json:"motivo_rechazo,omitempty"`

	Proceso          *EleccionProceso `gorm:"foreignKey:ProcesoID" json:"proceso,omitempty"`
	TitularAprendiz  *Aprendiz        `gorm:"foreignKey:TitularAprendizID" json:"titular_aprendiz,omitempty"`
	SuplenteAprendiz *Aprendiz        `gorm:"foreignKey:SuplenteAprendizID" json:"suplente_aprendiz,omitempty"`
}

func (EleccionPlancha) TableName() string { return "eleccion_planchas" }

// EleccionVoto registro transparente (quién votó y por qué plancha).
type EleccionVoto struct {
	BaseModel
	ProcesoID         uint `gorm:"column:proceso_id;not null;uniqueIndex:idx_eleccion_voto_proceso_user" json:"proceso_id"`
	PlanchaID         uint `gorm:"column:plancha_id;not null" json:"plancha_id"`
	VotanteUserID     uint `gorm:"column:votante_user_id;not null;uniqueIndex:idx_eleccion_voto_proceso_user" json:"votante_user_id"`
	VotanteAprendizID uint `gorm:"column:votante_aprendiz_id;not null" json:"votante_aprendiz_id"`

	Proceso         *EleccionProceso `gorm:"foreignKey:ProcesoID" json:"proceso,omitempty"`
	Plancha         *EleccionPlancha `gorm:"foreignKey:PlanchaID" json:"plancha,omitempty"`
	VotanteAprendiz *Aprendiz        `gorm:"foreignKey:VotanteAprendizID" json:"votante_aprendiz,omitempty"`
}

func (EleccionVoto) TableName() string { return "eleccion_votos" }

// EleccionResultado conteo y ganador del proceso.
type EleccionResultado struct {
	BaseModel
	ProcesoID         uint    `gorm:"column:proceso_id;not null;uniqueIndex" json:"proceso_id"`
	PlanchaGanadoraID *uint   `gorm:"column:plancha_ganadora_id" json:"plancha_ganadora_id,omitempty"`
	VotosTotales      int     `gorm:"column:votos_totales;not null;default:0" json:"votos_totales"`
	DetalleJSON       string  `gorm:"column:detalle_json;type:text" json:"detalle_json"`
	Empate            bool    `gorm:"column:empate;not null;default:false" json:"empate"`
	NotaDesempate     *string `gorm:"column:nota_desempate;type:text" json:"nota_desempate,omitempty"`
	UserRegistroID    *uint   `gorm:"column:user_registro_id" json:"user_registro_id,omitempty"`

	Proceso         *EleccionProceso `gorm:"foreignKey:ProcesoID" json:"proceso,omitempty"`
	PlanchaGanadora *EleccionPlancha `gorm:"foreignKey:PlanchaGanadoraID" json:"plancha_ganadora,omitempty"`
}

func (EleccionResultado) TableName() string { return "eleccion_resultados" }

// RepresentanteAprendiz vigencia histórica y actual por regional.
type RepresentanteAprendiz struct {
	BaseModel
	RegionalID           uint       `gorm:"column:regional_id;not null;index" json:"regional_id"`
	ProcesoID            uint       `gorm:"column:proceso_id;not null" json:"proceso_id"`
	TitularAprendizID    uint       `gorm:"column:titular_aprendiz_id;not null" json:"titular_aprendiz_id"`
	SuplenteAprendizID   uint       `gorm:"column:suplente_aprendiz_id;not null" json:"suplente_aprendiz_id"`
	VigenciaDesde        time.Time  `gorm:"column:vigencia_desde;not null" json:"vigencia_desde"`
	VigenciaHasta        *time.Time `gorm:"column:vigencia_hasta" json:"vigencia_hasta,omitempty"`

	Regional         *Regional        `gorm:"foreignKey:RegionalID" json:"regional,omitempty"`
	Proceso          *EleccionProceso `gorm:"foreignKey:ProcesoID" json:"proceso,omitempty"`
	TitularAprendiz  *Aprendiz        `gorm:"foreignKey:TitularAprendizID" json:"titular_aprendiz,omitempty"`
	SuplenteAprendiz *Aprendiz        `gorm:"foreignKey:SuplenteAprendizID" json:"suplente_aprendiz,omitempty"`
}

func (RepresentanteAprendiz) TableName() string { return "representantes_aprendiz" }

File: cdattg_web_frontend/src/types/eleccion.ts
export type EleccionProceso = {
  id: number;
  regional_id: number;
  regional_nombre?: string;
  anio: number;
  nombre_ciclo: string;
  estado: 'borrador' | 'inscripcion' | 'votacion' | 'empate_pendiente' | 'cerrada';
  fecha_inscripcion_inicio?: string;
  fecha_inscripcion_fin?: string;
  fecha_votacion_inicio?: string;
  fecha_votacion_fin?: string;
  min_dias_matricula?: number;
  planchas_confirmadas?: number;
  votos_registrados?: number;
  aprendices_elegibles?: number;
  created_at: string;
};

export type EleccionAprendizResumen = {
  id: number;
  nombre: string;
  ficha?: string;
  sede?: string;
};

export type EleccionPlancha = {
  id: number;
  proceso_id: number;
  estado: string;
  titular: EleccionAprendizResumen;
  suplente: EleccionAprendizResumen;
  titular_confirmado: boolean;
  suplente_confirmado: boolean;
  votos_recibidos?: number;
  motivo_rechazo?: string;
  pendiente_mi_confirmacion?: boolean;
};

export type EleccionResultado = {
  proceso_id: number;
  estado_proceso: string;
  plancha_ganadora_id?: number;
  votos_totales: number;
  participacion_pct: number;
  empate: boolean;
  nota_desempate?: string;
  conteo: { plancha_id: number; label: string; votos: number }[];
  votos?: {
    votante_nombre: string;
    votante_documento?: string;
    plancha_id: number;
    plancha_label: string;
    votado_at: string;
  }[];
};

export type RepresentanteAprendiz = {
  regional_id: number;
  regional_nombre?: string;
  proceso_id: number;
  nombre_ciclo?: string;
  anio?: number;
  titular: EleccionAprendizResumen;
  suplente: EleccionAprendizResumen;
  vigencia_desde: string;
  vigencia_hasta?: string | null;
};

export type EleccionMiRegional = {
  regional_id: number;
  regional_nombre?: string;
  proceso?: EleccionProceso;
  representantes_vigentes?: RepresentanteAprendiz;
  mi_aprendiz_id?: number;
  puede_votar: boolean;
  puede_postular: boolean;
  ya_voto?: boolean;
  es_candidato?: boolean;
  mi_voto_plancha_id?: number;
  planchas_pendientes_confirmar?: EleccionPlancha[];
};

export type EleccionProcesoRequest = {
  regional_id: number;
  anio: number;
  nombre_ciclo: string;
  fecha_inscripcion_inicio?: string | null;
  fecha_inscripcion_fin?: string | null;
  fecha_votacion_inicio?: string | null;
  fecha_votacion_fin?: string | null;
  min_dias_matricula?: number | null;
};

export type EleccionPlanchaRequest = {
  rol_candidatura: 'titular' | 'suplente';
  companero_aprendiz_id: number;
};

export type EleccionVotoRequest = {
  plancha_id: number;
};

export type EleccionDesempateRequest = {
  plancha_ganadora_id: number;
  nota_desempate: string;
};
